### ISSUE TITLE
Privacy Badger causes ad to show

### ISSUE THREAD

**Issue by terrorist96** [EFForg/privacybadgerchrome#626]
http://www.torrenthound.com/hash/10b0592b429b1392925c66a935031e058d17365f/torrent-info/linux-lite-2.4-64bit.iso

I have Adblock Plus and Privacy Badger (and other extensions, but they're irrelevant here).
If I have both extensions enabled, the green "Download Direct" ad link is present under the legitimate blue "Download .torrent" link. If I whitelist the site in Privacy Badger, the ad still shows. If I disable Privacy Badger via the extensions menu, ABP will be allowed to block the ad.

tl;dr: PB is stopping ABP from blocking an ad.

**Comment by SwartzCr** [comment]
Interesting, can you tell us what browser version you have, and which versions of Privacy Badger & ABP you are running?

**Comment by terrorist96** [comment]
Version 45.0.2454.101 m (64-bit)
ABP 1.9.3
PB  2015.8.14

As a side note, the extension Disconnect also caused this issue, but I uninstalled it and use Ghostery now (which doesn't have this issue).

More info:
https://forums.lanik.us/viewtopic.php?f=62&t=23134

**Comment by terrorist96** [comment]
Is this still being investigated?

**Comment by cooperq** [comment]
I don't think anyone has had a chance to investigate this yet, sorry. I suspect though that this is due to a quirk of how chrome works. We use the WebRequest API  to modify and cancel requests to blocked and cookieblocked domains in privacy badger. According to the chrome API documentation:
> Only one extension is allowed to redirect a request or modify a header at a time. If more than one extension attempts to modify the request, the most recently installed extension wins and all others are ignored. An extension is not notified if its instruction to modify or redirect has been ignored.
I suspect that this is what's causing the behavior that you are seeing, in which case there is nothing we can do about it.

**Comment by terrorist96** [comment]
But I don't understand how this is only an issue for this one specific site, and for blocking of this one specific ad only. In addition, whitelisting the site is ineffective. Privacy Badger has to be completely deactivated via the extensions menu. Shouldn't whitelisting/disabling PB on that one site not trigger the issue?

**Comment by terrorist96** [comment]
Hopefully you can add this to your list to look into eventually?

**Comment by terrorist96** [comment]
Privacy Badger also causes an ads to show on speedtest.net. See here:
https://forums.lanik.us/viewtopic.php?f=62&t=25113&p=85952#p85951

**Comment by cooperq** [comment]
I just tried to reproduce this on speedtest and on torrenthound and I am completely unable to reproduce it. @gorhill suggests that the issue might be some anti adblocking tech (disabling the hidden root stylesheets). Perhaps the problem is related to that? Either way, I don't see how privacy badger could be causing this. 
The only possibility that occurs to me is that possibly the other adblocker is just trying to hide ads with CSS and something that privacy badger is doing is removing those CSS overrides.

**Comment by terrorist96** [comment]
Regarding speedtest, you need to have the 3 custom filters added to ABP since those filters haven't been merged to EasyList yet.
Regarding Torrenthound, I just tried and it is still reproducible for me on my end using Chrome 48 (64bit). I have a separate profile with no modifications except for ABP installed. I loaded this:
http://www.torrenthound.com/hash/2ff489c4e35af05e11243836e75dc16a941270d7/torrent-info/The-Revenant-2015-DVDScr-XviD-AC3-ETRG
and the green DOWNLOAD DIRECT ad below DOWNLOAD .TORRENT is not visible. Then, I installed PB and refreshed the page and the DOWNLOAD DIRECT text ad is visible, even though PB isn't blocking anything on the page. 
![1](https://cloud.githubusercontent.com/assets/12802056/12867470/ce1c3cae-ccbc-11e5-95c7-bc51c9d03851.PNG)
![2](https://cloud.githubusercontent.com/assets/12802056/12867471/ce1d8a1e-ccbc-11e5-8149-d49e6b85c589.PNG)

**Comment by gorhill** [comment]
That's not Privacy Badger's fault, there is a timing issue going on in the page. If PB is there (or I suppose it could be any other extension which inject content scripts, the element to hide will end up having a style which override ABP's one. If ABP is alone, that style is overriden by ABP.

Summarily the element to hide has two competing CSS rules applied to it, having the same specificity, and sometimes the ABP wins, other time the page's own style wins. This can all be seen easily using the dev console:

![a](https://cloud.githubusercontent.com/assets/585534/12868831/3d902d00-ccdd-11e5-8245-0938fcddf21b.png)

**Comment by terrorist96** [comment]
So is there no solution? All I know is that the ad is blocked when only ABP is installed, but the ad is visible when ABP and PB are both installed, as demonstrated by my screenshots above.

I also tested by creating a new profile, installed PB first, loaded torrenthound, the ad was there (obviously), then I installed ABP, refreshed the page, ad was still there. Then disabled PB and refreshed the page and ad was gone. So it's unrelated to the order of extensions installed, as mentioned originally by Cooper. 

> If more than one extension attempts to modify the request, the most recently installed extension wins and all others are ignored

**Comment by gorhill** [comment]
> So is there no solution?

Well it sort of work with AdBlock, they also declare `visibility: hidden;` in their CSS rule, so the button is not visible but on the other hand the placeholder is not collapsed for the current case -- a lesser issue I suppose.

**Comment by cooperq** [comment]
Thanks for the explanation @gorhill ! Closing this issue for now since there is nothing we can do about it.

**Comment by terrorist96** [comment]
I didn't think he said there was nothing that could be done about it. Just that it's a tricky problem. Please leave this open.

**Comment by gorhill** [comment]
@terrorist96 it's not a Privacy Badger issue, it's an ABP issue.

**Comment by terrorist96** [comment]
Okay, thanks. I mentioned it here:
https://forums.lanik.us/viewtopic.php?f=62&t=23134&start=15

Hopefully someone from the ABP team sees it.

**Comment by gorhill** [comment]
##### With Privacy Badger ("PB"):

1. ABP injects its CSS rules
1. Page load its CSS rules (from `layout.css?v=[...]`) 

Result: the page's own rule wins.

##### Without Privacy Badger:

1. Page load its CSS rules (from `layout.css?v=[...]`) 
1. ABP injects its CSS rules

Result: ABP's CSS rule wins.

##### Conclusion

PB is an innocent bystander here, the real issue is one of timing. ABP is sensitive from having the injection of its CSS rules being delayed -- this sensitivity is not PB's fault, it could be any other extension(s) than PB.
